sqlmesh/core/state_sync/engine_adapter.py

```diff
@@ -369,6 +409,68 @@ def _get_environment(
         env = self._environment_from_row(row)
         return env
 
+    @transactional()
+    def migrate(self) -> None:
+        super().migrate()
+
+    @transactional()
+    def _migrate_rows(self) -> None:
+        all_snapshots = self._get_snapshots(lock_for_update=True)
+        environments = self.get_environments()
+        snapshots_by_version = defaultdict(list)
+
+        for snapshot in all_snapshots.values():
+            snapshots_by_version[(snapshot.name, snapshot.version)].append(snapshot)
+
+        new_snapshots = set()
+
+        for environment in environments:
+            snapshots = []
+            models = {}
+            audits = {}
+            cache: t.Dict[str, SnapshotFingerprint] = {}
+
+            for info in environment.snapshots:
+                snapshot = all_snapshots[info.snapshot_id]
+                snapshots.append(snapshot)
+                model = snapshot.model
+                models[model.name] = model
+                for audit in snapshot.audits:
+                    audits[audit.name] = audit
+
+            environment.snapshots.clear()
+
+            for snapshot in snapshots:
+                model = snapshot.model
+                snapshot = Snapshot.from_model(
+                    snapshot.model,
+                    physical_schema=snapshot.physical_schema,
+                    models=models,
+                    ttl=snapshot.ttl,
+                    version=snapshot.version,
+                    audits=audits,
+                    cache=cache,
+                )
+
+                environment.snapshots.append(snapshot.table_info)
+
+                if snapshot in new_snapshots:
+                    continue
+
+                for other in snapshots_by_version[(snapshot.name, snapshot.version)]:
+                    if snapshot is not other:
+                        snapshot.merge_intervals(other)
+
+                new_snapshots.add(snapshot)
```

Comment:
Should we check whether fingerprint has changed instead of always doing an unconditional overwrite?